runtime-modules/content/src/vnft_auction/types.rs

```diff
@@ -0,0 +1,279 @@
+    pub fn ensure_ownership<T: Trait>(&self, owner: &AccountId) -> DispatchResult {
+        ensure!(self.owner.eq(owner), Error::<T>::DoesNotOwnVNFT);
+        Ok(())
+    }
+}
+
+/// Owned vNFT representation
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, Debug)]
+pub struct OwnedNFT<
+    AccountId,
+    Moment: BaseArithmetic + Copy,
+    CuratorGroupId: Default + Copy,
+    CuratorId: Default + Copy,
+    MemberId: Default + Copy,
+    Balance,
+> {
+    pub owner: AccountId,
+    pub transactional_status:
+        TransactionalStatus<AccountId, Moment, CuratorGroupId, CuratorId, MemberId, Balance>,
+    pub creator_royalty: Option<(AccountId, Royalty)>,
+}
+
+impl<
+        AccountId: PartialEq,
+        Moment: BaseArithmetic + Copy,
+        CuratorGroupId: Default + Copy,
+        CuratorId: Default + Copy,
+        MemberId: Default + Copy,
+        Balance,
+    > OwnedNFT<AccountId, Moment, CuratorGroupId, CuratorId, MemberId, Balance>
+{
+    pub fn is_owner(&self, account_id: &AccountId) -> bool {
+        self.owner.eq(account_id)
+    }
+}
+
+/// Enum, representing nft issuance status
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
+pub enum NFTStatus<
+    AccountId,
+    Moment: BaseArithmetic + Copy,
+    CuratorGroupId: Default + Copy,
+    CuratorId: Default + Copy,
+    MemberId: Default + Copy,
+    Balance,
+> {
+    NoneIssued,
+    Owned(OwnedNFT<AccountId, Moment, CuratorGroupId, CuratorId, MemberId, Balance>),
+}
+
+impl<
+        AccountId,
+        Moment: BaseArithmetic + Copy,
+        CuratorGroupId: Default + Copy,
+        CuratorId: Default + Copy,
+        MemberId: Default + Copy,
+        Balance,
+    > Default for NFTStatus<AccountId, Moment, CuratorGroupId, CuratorId, MemberId, Balance>
+{
+    fn default() -> Self {
+        Self::NoneIssued
+    }
+}
+
+/// Either new auction, which requires vNFT issance or auction for already existing nft.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
+pub enum AuctionMode {
+    // Auction, where nft issued at the end
+    WithIssuance(Option<Royalty>, Metadata),
+    // Auction for already existing nft
+    WithoutIsuance,
+}
+
+impl Default for AuctionMode {
+    fn default() -> Self {
+        Self::WithoutIsuance
+    }
+}
+
+/// Information on the auction being created.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, Debug)]
+pub struct AuctionRecord<
+    AccountId,
+    Moment: BaseArithmetic + Copy,
+    CuratorGroupId: Default + Copy,
+    CuratorId: Default + Copy,
+    MemberId: Default + Copy,
+    Balance,
+> {
+    pub auctioneer: ContentActor<CuratorGroupId, CuratorId, MemberId>,
+    pub auctioneer_account_id: AccountId,
+    pub auction_mode: AuctionMode,
```

Comment:
I still don't understand why we need this?

Reply:
When we are going to start auction for nft, there are 2 options:
- start auction for already existing nft
- start auction for where previously non-existing nft would be released by the end of auction. 

To be able to start auction for previously non-existing nft, we need to provide a metadata ( which would be raised in the event) and optional creator royalty (which would be stored) through the extrinsic parameters.
Afterwards we store this data inside the `AuctionRecord` structure and release it after it was completed. 
It results in setting royalty parameter and raising metadata in the event during nft issuance.
```r
/// Issue vnft
    pub(crate) fn issue_vnft(
        video: &mut Video<T>,
        video_id: T::VideoId,
        owner: T::AccountId,
        creator_royalty: Option<Royalty>,
    ) {
        video.nft_status = NFTStatus::Owned(OwnedNFT {
            owner,
            transactional_status: TransactionalStatus::Idle,
            creator_royalty,
        });

        Self::deposit_event(RawEvent::NftIssued(video_id, video.nft_status.clone()));
    }
```

Reply:
Remember
- Metadata should not be stored in storage, so that field goes away.
- Royalty is not set per auction, it is set per NFT when created, so this value should live as an `Option<Royalty>` in `OwnedNFT`.

When these to fields are gone, the only role of `AuctionMode` is to distinguished between whether this auction is the first one, or not, but I don't think the runtime needs to remember this, does.

> To be able to start auction for previously non-existing nft, we need to provide a metadata ( which would be raised in the event) and optional creator royalty (which would be stored) through the extrinsic parameters.

If we want to allow issuing an NFT & opening an auction, in one step, then we should make a dedicated extrinsic for this combination, or to make the parameters for `issue` richer, so that one can add auction creation parameters optionally.

Reply:
> Remember
> 
> * Metadata should not be stored in storage, so that field goes away.
> * Royalty is not set per auction, it is set per NFT when created, so this value should live as an `Option<Royalty>` in `OwnedNFT`.
> 
> When these to fields are gone, the only role of `AuctionMode` is to distinguished between whether this auction is the first one, or not, but I don't think the runtime needs to remember this, does.
> 
> > To be able to start auction for previously non-existing nft, we need to provide a metadata ( which would be raised in the event) and optional creator royalty (which would be stored) through the extrinsic parameters.
> 
> If we want to allow issuing an NFT & opening an auction, in one step, then we should make a dedicated extrinsic for this combination, or to make the parameters for `issue` richer, so that one can add auction creation parameters optionally.

Agree with the second point here, we can put `Option<Royalty>` straight into the `OwnedNFT` representation at the auction start. 

To begin with, nft can either be issued instantly at no cost or at the time of auction completion, if we run auction for non existing vnft. Auction can be run for both already existing nfts and non-existing nft, which would be issued after auction completion.

We should only raise metadata at the time of nft issuance, not at the moment of auction start, so we need a way to temporary store that value, don`t we?

Reply:
> To begin with, nft can either be issued instantly at no cost or at the time of auction completion, if we run auction for non existing vnft. Auction can be run for both already existing nfts and non-existing nft, which would be issued after auction completion.

I just tihnk of the NFt as issued the moment you create it, not the moment the auction is completed, but I am not sure this semantic difference has much significance to it.

> We should only raise metadata at the time of nft issuance, not at the moment of auction start, so we need a way to temporary store that value until the respective event wouldn't be triggered, don`t we?

The metadata has to exist from the very beginning, before any auction, because it is this metadata which will be used by the user facing apps to display a description of the NFT, perhaps some different cover photo, etc. So think of it the same way as the metadata of a video, that has to exist out of the gate when the video is created in the channel.

---

If we are not 100% synched after this reply, lets jump on a quick call, will be easier ;)

Reply:
refactored it here https://github.com/Joystream/joystream/pull/2512/commits/638282dd2461f817e351f0d8ca4c90c33d388d1a